### PR TITLE
ILSGateway: Allow users to self-register

### PR DESCRIPTION
**Pull request by gcapalbo**
The first 3 commits update the messaging framework to allow sms to be tied to a domain when they are sent to a domain-owned gateway from an unregistered contact.

The rest of the PR updates the ILSGateway sms code to allow unregistered users to self-register.

Best reviewed commit by commit
@NoahCarnahan

### CODE REVIEW

File: custom/ilsgateway/tanzania/handlers/register.py [hunk 1]
```diff
@@ -92,9 +117,20 @@ def handle(self):
         elif split_name:
             first_name = split_name[0]
 
+        if not self.user:
+            key = u'generating ils username for %s, %s, %s' % (self.domain, first_name, last_name)
+            with CriticalSection([key]):
+                username = generate_username(self.domain, first_name, last_name)
+                password = uuid.uuid4().hex
+                self.user = CommCareUser.create(self.domain, username, password, phone_number=strip_plus(self.msg.phone_number))
```
Comment (dimagimon) [inline custom/ilsgateway/tanzania/handlers/register.py]: E501 line too long (128 > 115 characters)

**Comment by NoahCarnahan** [comment]
Test failures look real, otherwise looks good.

**Comment by gcapalbo** [comment]
yep, fixing those now, thanks

**Comment by gcapalbo** [comment]
:white_check_mark:

**Pull request merged**